tahoe_sites/backends.py

```diff
@@ -0,0 +1,62 @@
+"""
+Authentication backends for Tahoe's multi-tenancy.
+"""
+from django.conf import settings
+from django.contrib.auth.backends import AllowAllUsersModelBackend, ModelBackend
+
+from tahoe_sites.api import get_current_site, get_organization_by_site, get_organizations_for_user
+
+
+class DefaultSiteBackend(ModelBackend):
+    """
+    User can log in to the default/root site (edx.appsembler.com) because it is required during the signup.
+    Also, superusers (appsembler admins) can log into any site.
+    """
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        """
+        Authenticate superusers only.
+        """
+        user = super().authenticate(request, username, password, **kwargs)
+
+        if user:
+            if user.is_superuser or get_current_site().id == settings.SITE_ID:
+                return user
+        return None
+
+
+class OrganizationMemberBackend(AllowAllUsersModelBackend):
+    """Backend for organization based authentication
+
+    This class is an extension of Django's `AllowAllUserModelBackend`
+
+    This class checks that the user to authenticate belongs to one of the
+    organizations in the specified site
+
+    This class extends `AllowAllUserModelBackend` instead of `ModelBackend`
+    because users need to be able to authenticate when the `user.is_active` is
+    `False`. The reason for this is that Open edX has an email verification
+    scheme that uses `User.is_active` in order to prevent user activity until
+    the users have verified their email. Effectively, edx-platform LMS user
+    authentication state is used to manage authorization state.
+
+    The key problem is that Django's `django.contrib.auth.backends.ModelBackend`
+    is called with `not user.is_active` and Ironwood introduced a check to test
+    authentication on a not yet authorized user. This breaks Tahoe multisite
+    behavior. Extending `AllowAllUsersModelBackend` restores correct behavior.
+
+    For further reference, see settings files and read Django documentation for
+    `AUTHENTICATION_BACKENDS`
+    """
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        """
+        Authenticate organization learners.
+        """
+        user = super().authenticate(request, username, password, **kwargs)
+
+        site = get_current_site()
+        is_default_site = site.id == settings.SITE_ID
+        if not is_default_site and user and not user.is_superuser:
+            user_organizations = get_organizations_for_user(user=user, with_inactive_users=True)
```

Comment:
**Required change:** I'm arguing here that `get_organizations_for_user` shouldn't be used and should be removed from this package here's why:

 - If a user has more than one organization, this is a data problem that should be fixed. 
 - If a user has more than one organization, few parts of our system will break.
 - We don't support this use case anymore and we're implementing one user multi-site at the Identity Provider level.

```suggestion
            user_organization = get_organization_for_user(user=user)
```


### Question about `with_inactive_users`
One more question, why the `with_inactive_users` is set to `True`?

Reply:
> One more question, why the with_inactive_users is set to True?

@OmarIthawi its because the backend business logic says so in the [original backend](https://github.com/appsembler/edx-organizations/blob/3708d65456e70328383fecae03bb48666f20e405/organizations/backends.py#L51)

also the original code is fetching regardless of being active or not
```python
user_organizations = set(user.organizations.values_list('short_name', flat=True))
```

Reply:
@OmarIthawi I didn't change this because it is the purpose of this backend as stated [here](https://github.com/appsembler/edx-organizations/blob/3708d65456e70328383fecae03bb48666f20e405/organizations/backends.py#L44) (related to email verification cycle). Please review, or let me know if this logic is really not needed